### PR TITLE
Config: Improve the node.config.php transformation

### PR DESCRIPTION
**Pull request by nupplaphil**
- Add more types
- Improvement for assoziative arrays and key-value arrays
- Add a lot more tests

### CODE REVIEW

**Comment by nupplaphil** [comment]
plz skip the license check .. *gg*

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 1]
```diff
@@ -26,67 +26,191 @@
  */
 class ConfigFileTransformer
 {
+	/**
+	 * The public method to start the encoding
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]: This doesn't say anything that the prototype of the method already says. What more can you say about this method? When is it supposed to be used? Caveats?

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 2]
```diff
@@ -26,67 +26,191 @@
...
+	}
 
-				foreach ($keys as $key) {
-					$dataString .= static::mapConfigValue($key, $data[$category][$key]);
-				}
-			}
-			$dataString .= "\t]," . PHP_EOL;
+	/**
+	 * Extracts an inner config array.
+	 * Either as a Key => Value pair array or as an assoziative array
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]: In English, associative == key => value.
```suggestion
	 * Either as an associative array or as a list
```

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 3]
```diff
@@ -26,67 +26,191 @@
...
+	 *    ...
+	 * ]
+	 *
+	 * @param array $config             The key-value array
+	 * @param int   $level              The current level of recursion (necessary for tab-indentation calculation)
+	 * @param bool  $inAssoziativeArray If true, the current array resides inside another assoziative array. Different rules may be applicable
+	 *
+	 * @return string The config string
+	 */
+	protected static function extractKeyValueArray(array $config, int $level = 0, bool $inAssoziativeArray = false): string
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]:
```suggestion
	protected static function extractAssociativeArray(array $config, int $level = 0, bool $inList = false): string
```

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 4]
```diff
@@ -26,67 +26,191 @@
...
+	 *    ],
+	 * ]
+	 *
+	 * @param array $config             The assoziative array
+	 * @param int   $level              The current level of recursion (necessary for tab-indentation calculation)
+	 * @param bool  $inAssoziativeArray If true, the current array resides inside another assoziative array. Different rules may be applicable
+	 *
+	 * @return string The config string
+	 */
+	protected static function extractAssoziativeArray(array $config, int $level = 0, bool $inAssoziativeArray = false): string
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]:
```suggestion
	protected static function extractList(array $config, int $level = 0, bool $inList = false): string
```

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 5]
```diff
@@ -26,67 +26,191 @@
...
 		}
 
+		$string .= str_repeat("\t", $level) . ']';
+
 		return $string;
 	}
 
-	protected static function mapConfigValue(string $key, $value, $level = 0): string
+	/**
+	 * Extracts an assoziative array and save it into a string
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]:
```suggestion
	 * Extracts a list and save it into a string
```

File: src/Core/Config/Util/ConfigFileTransformer.php [hunk 6]
```diff
@@ -26,67 +26,191 @@
...
+	/**
+	 * Transforms one config value and returns the corresponding text-representation
+	 *
+	 * @param mixed $value              Any value to transform
+	 * @param int   $level              The current level of recursion (necessary for tab-indentation calculation)
+	 * @param bool  $inAssoziativeArray If true, the current array resides inside another assoziative array. Different rules may be applicable
+	 *
+	 * @return string
+	 */
+	protected static function transformConfigValue($value, int $level = 0, bool $inAssoziativeArray = false): string
```
Comment (MrPetovan) [inline src/Core/Config/Util/ConfigFileTransformer.php]:
```suggestion
	protected static function transformConfigValue($value, int $level = 0, bool $inList = false): string
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**